pkg/sfsturbo/nodeserver.go

```diff
@@ -17,142 +17,223 @@ limitations under the License.
+	if !isMounted(targetPath) {
+		log.Infof("NodeUnpublishVolume: %s has already uMounted", targetPath)
+		return &csi.NodeUnpublishVolumeResponse{}, nil
 	}
-	klog.V(2).Infof("NodeUnpublishVolume: unmount volume %s on %s successfully", volumeID, targetPath)
 
+	if err := Unmount(targetPath); err != nil {
+		return nil, status.Errorf(codes.Internal, "Failed to unmount target %q: %v", targetPath, err)
+	}
+	log.Infof("NodeUnpublishVolume: unmount volume %s on %s successfully", volumeID, targetPath)
 	return &csi.NodeUnpublishVolumeResponse{}, nil
 }
 
-func (ns *nodeServer) NodeGetInfo(ctx context.Context, req *csi.NodeGetInfoRequest) (*csi.NodeGetInfoResponse, error) {
-	klog.V(2).Infof("NodeGetInfo called with request %v", *req)
+func (ns *nodeServer) NodeGetInfo(_ context.Context, req *csi.NodeGetInfoRequest) (*csi.NodeGetInfoResponse, error) {
+	log.Infof("NodeGetInfo: called with args %v", protosanitizer.StripSecrets(*req))
+	nodeID, err := ns.Metadata.GetInstanceID()
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Unable to retrieve instance id of node %s", err)
+	}
+
+	zone, err := ns.Metadata.GetAvailabilityZone()
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Unable to retrieve availability zone of node %v", err)
+	}
+	topology := &csi.Topology{Segments: map[string]string{topologyKey: zone}}
+	log.Infof("NodeGetInfo nodeID: %s, topology: %s", nodeID, protosanitizer.StripSecrets(*topology))
 	return &csi.NodeGetInfoResponse{
-		NodeId: ns.Driver.nodeID,
+		NodeId:             nodeID,
+		AccessibleTopology: topology,
 	}, nil
 }
 
-func (ns *nodeServer) NodeGetCapabilities(ctx context.Context, req *csi.NodeGetCapabilitiesRequest) (*csi.NodeGetCapabilitiesResponse, error) {
-	klog.V(2).Infof("NodeGetCapabilities called with req: %#v", req)
-
+func (ns *nodeServer) NodeGetCapabilities(_ context.Context, req *csi.NodeGetCapabilitiesRequest) (
+	*csi.NodeGetCapabilitiesResponse, error) {
+	log.Infof("NodeGetCapabilities: called with args %v", protosanitizer.StripSecrets(*req))
 	return &csi.NodeGetCapabilitiesResponse{
 		Capabilities: ns.Driver.nscap,
 	}, nil
 }
 
-func (ns *nodeServer) NodeGetVolumeStats(_ context.Context, req *csi.NodeGetVolumeStatsRequest) (*csi.NodeGetVolumeStatsResponse, error) {
-	return nil, status.Error(codes.Unimplemented, "")
+func (ns *nodeServer) NodeGetVolumeStats(_ context.Context, req *csi.NodeGetVolumeStatsRequest) (
+	*csi.NodeGetVolumeStatsResponse, error) {
+	log.Infof("NodeGetVolumeStats: called with args %v", protosanitizer.StripSecrets(*req))
+
+	volumeID := req.GetVolumeId()
+	volumePath := req.GetVolumePath()
+	if err := nodeGetStatsValidation(volumeID, volumePath); err != nil {
+		return nil, err
+	}
+
+	stats, err := ns.Mount.GetDeviceStats(volumePath)
+	if err != nil {
+		return nil, status.Errorf(codes.Unknown, "Failed to get stats by path: %s", err)
+	}
+	log.Infof("NodeGetVolumeStats: stats info :%s", protosanitizer.StripSecrets(*stats))
+	if stats.Block {
```

Comment:
Does SFS Turbo support block file systems?

Reply:
it does not support block